Removed collision object count limits on maps

only-a-ptr opened the pull request:
I think I nailed it. The messy code in 'Collisions.h/cpp' made it hard to see what's going on (I'm looking at you, estama!), but I figured it out. I quickly tested by driving Desperado buggy on Auriga - everything seemed to work just fine.

## Changes
* The collision box limit is now 1 million
* The collision triangle limit is now 4 billion
* The memory is dynamically sized - you can forget the 'collision-tris' entry from .terrn2
* The `COLL: The hashtable is full` error is history.

Fixes #1395

tritonas00 commented:
Compiles fine, no crashes so far, :+1:  from me

Michael10055 commented:
Hm, doesn't seem to work from my tests. A while back I tried merging both the original NeoQ and NeoQ2.0 into one map just to see if it was possible, it goes over the collision limit. This change doesn't really seem to make a difference - spams the log with  "hashtable full" errors after a certain point.
````
15:59:34: Mesh: Loading NeoQ2-0highway-section-AColisionante.mesh.
15:59:34: WARNING: NeoQ2-0highway-section-AColisionante.mesh is an older format ([MeshSerializer_v1.40]); you should upgrade it as soon as possible using the OgreMeshUpgrade tool.
15:59:34: Can't assign material ModeloColisionante to SubEntity of Ogre/MO1099 because this Material does not exist. Have you forgotten to define it in a .material script?
15:59:37: COLL: The hashtable is full.
```
And both maps definitely don't equal over 4 billion 😆

only-a-ptr commented:
@Michael10055 I checked and things are worse than I thought. Apparently this message has nothing to do with triangle/box count, it relates only to the internal hashtable. The problem is that it's not a regular hashtable but some pseudo-sparse hybrid, and it's aparently broken.

As I already said in #1395, analyzing what's wrong here isn't worth the work - the whole RoR static collision system would best be thrown away and replaced by collision detection facilities from some proven physics engine - I favor Bullet which explicitly offers this in the docs.

But I'll make one last try - I'll try to reduce the lookup system to an ordinary hashtable which can handle overflows, so it doesn't get "full".

DarthCainRor commented:
I'm going to post a test branch of the community map later that will cause this issue for testing.

DarthCainRor commented:
https://github.com/RigsOfRods-Community/community-map/tree/Collision-breaker

only-a-ptr commented:
I pushed a new update. I can't test with the community map right now (at work), I just quickly drove Desperado on Auriga again - collisions seemed just fine.

only-a-ptr commented:
I created a PR for the 'pointlight crash' (https://github.com/RigsOfRods/rigs-of-rods/pull/1398) and also included it here to test (will be excluded in the end).

Then I tested the CommunityMap 'collision breaker' edition... didn't break me!
![screenshot_2017-10-05_18-13-34_1](https://user-images.githubusercontent.com/491088/31238232-a3d8bc94-a9f9-11e7-855e-1db54ea162da.jpg)

Michael10055 commented:
Seems to work fine now 👍 
![image](https://user-images.githubusercontent.com/11002490/31255071-16202950-a9f9-11e7-8133-2a87f7b03f76.png)
The test map I mentioned in a earlier comment works now as well.

only-a-ptr commented:
@Michael10055 I'm thrilled that it's working. It may not look like a miracle on the surface, but as programming goes, I really surprised myself here. The Collisions code is dreadful, littered with ambiguous (dangerous!) programming constructs and downright ugly writing (I hate you, Estama! 😆  ). For the longest time, I was simply avoiding it, so it's relieving to know I can actually work with it.

Those `COL:` stats in RoR.log were probably ever meant for a developer, for example the `Cell size:` line just prints a constant value. I already removed couple lines from there. And the `hashtable collisions` is dead now, I removed the (ugly!) counting code.
